client/src/client/session_manager_client.py

```diff
@@ -154,16 +146,42 @@ async def _handle_upstream(self, conn: ws.WebSocketClientProtocol) -> None:
             message = await self.upstream_message_queue.get()
             receiver = message.header.receiver
 
-            # Currently receiver being None means the message should be
-            # terminated by the server
-            if receiver is not None:
-                if receiver in self.potential_sessions.keys():
-                    self._activate_outbound_session(receiver)
-
-                if receiver in self.active_sessions.keys():
-                    message = self._encrypt_user_payload(message)
-
-            await cans_send(message, conn)
+            if receiver is None:
+                # Receiver is None - the message should be terminated by
+                # the server and so no crypto session is used
+                await cans_send(message, conn)
+            elif receiver in self.state_machine.potential_sessions.keys():
```

Comment:
maybe some nice method here? Getting keys for something (it's not obvious it's a dictionary) looks ambiguous

Reply:
After reading sessions_state_machine.py I am not so sure about it though